File: Trees/110. Balanced Binary Tree/Solution.java
import java.lang.Math;

//Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

public class Solution {
    Boolean isBalance = true;
    private int dfs(TreeNode root) {
        if (root == null) return 0;

        int left = dfs(root.left);
        int right = dfs(root.right);

        if (Math.abs(left - right) > 1) isBalance = false;

        return 1 + Math.max(left, right);
    }
    public boolean isBalanced(TreeNode root) {
        // N : no of nodes in a tree
        // Time Complexity: O(N)
        // Space Complexity: O(N) -> Stack Space
        dfs(root);
        return isBalance;
    }
}

File: Trees/110. Balanced Binary Tree/Solution.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        # N : no of nodes in a tree
        # Time Complexity: O(N)
        # Space Complexity: O(N) -> Stack Space
        isBalance = True
        def dfs(root):
            nonlocal isBalance
            if not root:
                return 0
            
            left = dfs(root.left)
            right = dfs(root.right)

            if (abs(left - right) > 1):
                isBalance = False

            return 1 + max(left, right)
        
        dfs(root)
        return isBalance